port/raspberrypi/rp2xxx/src/hal/atomic.zig

```diff
@@ -15,60 +23,193 @@ inline fn atomic_unlock(critical_section: CriticalSection) void {
     const save = atomic_lock();
     defer atomic_unlock(save);
     const tmp = ptr.*;
-    ptr.* = tmp & val;
+
+    switch (op) {
+        .Xchg => ptr.* = val,
+        .Add => ptr.* += val,
+        .Sub => ptr.* -= val,
+        .And => ptr.* &= val,
+        .Or => ptr.* |= val,
+        .Xor => ptr.* ^= val,
+        .Nand => ptr.* = ~(ptr.* & val),
+        .Max => ptr.* = @max(ptr.*, val),
+        .Min => ptr.* = @min(ptr.*, val),
+    }
+
     return tmp;
 }
 
-fn atomic_rmw_or(comptime T: type, ptr: *volatile T, val: T, _: i32) T {
+fn atomic_compare_exchange(comptime T: type, ptr: *volatile T, expected: *T, desired: T, _: i32, _: i32) bool {
     const save = atomic_lock();
     defer atomic_unlock(save);
-    const tmp = ptr.*;
-    ptr.* = tmp | val;
-    return tmp;
+    const old_value = ptr.*;
+
+    if (old_value == expected.*) {
+        ptr.* = desired;
+        return true;
+    } else {
+        expected.* = old_value;
+        return false;
+    }
 }
 
-fn __atomic_fetch_and_4(ptr: *u32, val: u32, model: i32) callconv(.c) u32 {
-    return atomic_rmw_and(u32, ptr, val, model);
+export fn __atomic_store_1(ptr: *u8, val: u8, model: i32) callconv(.c) void {
+    atomic_store(u8, ptr, val, model);
 }
 
-fn __atomic_fetch_or_4(ptr: *u32, val: u32, model: i32) callconv(.c) u32 {
-    return atomic_rmw_or(u32, ptr, val, model);
+export fn __atomic_store_2(ptr: *u16, val: u16, model: i32) callconv(.c) void {
+    atomic_store(u16, ptr, val, model);
 }
 
-fn __atomic_load_4(src: *u32, model: i32) callconv(.c) u32 {
-    return atomic_load(u32, src, model);
+export fn __atomic_store_4(ptr: *u32, val: u32, model: i32) callconv(.c) void {
+    atomic_store(u32, ptr, val, model);
 }
 
-const linkage: std.builtin.GlobalLinkage = if (builtin.is_test)
-    .internal
-else if (builtin.object_format == .c)
-    .strong
-else
-    .weak;
+export fn __atomic_load_1(ptr: *u8, model: i32) callconv(.c) u8 {
+    return atomic_load(u8, ptr, model);
+}
 
-const visibility: std.builtin.SymbolVisibility = .default;
+export fn __atomic_load_2(ptr: *u16, model: i32) callconv(.c) u16 {
+    return atomic_load(u16, ptr, model);
+}
 
-// Based on: https://github.com/ziglang/zig/blob/79460d4a3eef8eb927b02a7eda8bc9999a766672/lib/compiler_rt/atomics.zig
-// and: https://github.com/raspberrypi/pico-sdk/blob/ee68c78d0afae2b69c03ae1a72bf5cc267a2d94c/src/rp2_common/pico_atomic/atomic.c
-// TODO: export all the rest atomics
+export fn __atomic_load_4(ptr: *u32, model: i32) callconv(.c) u32 {
+    return atomic_load(u32, ptr, model);
+}
 
-comptime {
-    // This export should only happen for the RP2040 due to the ARMv6-M ISA used by the ARM Cortex-M0+.
-    if (chip == .RP2040) {
-        @export(&__atomic_fetch_and_4, .{ .name = "__atomic_fetch_and_4", .linkage = linkage, .visibility = visibility });
-        @export(&__atomic_fetch_or_4, .{ .name = "__atomic_fetch_or_4", .linkage = linkage, .visibility = visibility });
```

Comment:
What about this note regarding the rp2040? Should these functions be exported for the rp2350?

Reply:
@mattnite RP2350 uses newer ARM ISA (ARMv8) where all those operations are dedicated instructions so there is no need to emulate atomics this way

Reply:
Yes, good question. RP2350 don't need those symbols as all those operations are made by dedicated instructions.